ipa-core/src/protocol/ipa_prf/shuffle/sharded.rs

```diff
@@ -0,0 +1,522 @@
+                    let mask = match direction {
+                        Direction::Left => l,
+                        Direction::Right => r,
+                    };
+
+                    item + mask
+                }),
+                |ctx, record_id, _| ctx.pick_shard(record_id, direction),
+            ))
+            .await?;
+
+            let ctx = self.narrow(&ShuffleStep::LocalShuffle);
+            resharded.shuffle(&mut match direction {
+                Direction::Left => ctx.prss_rng().0,
+                Direction::Right => ctx.prss_rng().1,
+            });
+
+            Ok(resharded)
+        }
+    }
+
+    /// Receive all the values from the specified helper. It is assumed that this context is narrowed
+    /// to the correct step, before receiving.
+    fn recv_all<S>(
+        &self,
+        direction: Direction,
+    ) -> impl Future<Output = Result<Vec<S>, crate::error::Error>> + Send
+    where
+        S: ShuffleShare,
+    {
+        async move {
+            let mut rid = RecordId::FIRST;
+            let mut buf = Vec::new();
+            let recv_channel = self.recv_channel(self.role().peer(direction));
+            loop {
+                match recv_channel.receive(rid).await {
+                    Ok(v) => buf.push(v),
+                    Err(Error::EndOfStream { .. }) => break,
+                    Err(e) => return Err(e.into()),
+                }
+
+                rid += 1;
+            }
+
+            Ok(buf)
+        }
+    }
+
+    /// Send all values to the specified helper. It is assumed that this context is narrowed
+    /// to the correct step, before receiving.
+    ///
+    /// It works for empty input too - this shard will open a connection and then immediately close
+    /// it, sending a signal to the receiver that the stream has been closed.
+    fn send_all<I, S>(
+        &self,
+        shares: I,
+        direction: Direction,
+    ) -> impl Future<Output = Result<(), crate::error::Error>>
+    where
+        I: IntoIterator<Item = S>,
+        I::IntoIter: ExactSizeIterator + Send,
+        S: ShuffleShare,
+    {
+        let shares = shares.into_iter();
+        let sz = shares.len();
+        let ctx = self.set_total_records(max(1, sz));
+
+        async move {
+            let send_channel = ctx.send_channel::<S>(ctx.role().peer(direction));
+
+            let mut send_stream = seq_join(
+                self.active_work(),
+                stream::iter(shares)
+                    .enumerate()
+                    .map(|(i, v)| send_channel.send(RecordId::from(i), v)),
+            );
+            while let Some(v) = send_stream.next().await {
+                v?;
+            }
+
+            send_channel.close(RecordId::from(sz)).await;
+
+            Ok(())
+        }
+    }
+}
+
+#[derive(Step)]
+enum ShuffleStep {
+    /// Depending on the helper position inside the MPC ring, generate Ã, B̃ or both.
```

Comment:
see my answer below - there seem to be an issue finding a publicly available version of this paper. I asked Daniel if he has one. Alternatively we can put a paywalled link but it will unlikely be useful for the majority of readers here.